Validate `authToken` from user data file before returning it

fredrikekelund opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

## Related issues

- Fixes https://github.com/Automattic/dotcom-forge/issues/10215
- Fixes https://github.com/Automattic/dotcom-forge/issues/10258

## Proposed Changes

In https://github.com/Automattic/dotcom-forge/issues/10215, we observed that the `authToken` object in the `appdata-v1.json` file sometimes misses the `id` prop. Even if the TS type previously designated this prop as optional, that's not actually how the code works—we need the WordPress.com user ID for several features that interact with the WP.com APIs.

In this PR, I've addressed this problem by:

1. Adding [zod](https://zod.dev/) as a dependency (a validation library with incredibly good TS support that's successfully being used on Tumblr)
2. Adding a zod schema for the `authToken` object in the user data file. Most properties are _not_ optional.
3. Making `getToken` in `src/lib/oauth.ts` return null if the zod validation fails.

Moreover, I took the opportunity to move the `onOpenUrlCallback` function from `src/index.ts` to `src/lib/oauth.ts`. Given it deals exclusively with OAuth-related logic, it makes more sense for it to live there. I also inlined the logic from `setUpAuthCallbackHandler` into `onOpenUrlCallback`, reducing the number of events being passed around.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Be logged out
2. Log in
3. Ensure that the Assistant works as expected
4. Ensure that connecting sites in the Sync tab works as expected
5. Ensure that the Share tab works as expected

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

fredrikekelund committed:
WIP

fredrikekelund committed:
Simplify logic

fredrikekelund committed:
Merge branch 'trunk' into f26d/validate-auth-token

fredrikekelund commented on src/tests/index.test.ts:
I had to edit this file to accommodate my changes to `src/lib/oauth.ts`, and I took the opportunity to simplify some previously somewhat convoluted logic.

ashfame commented on src/lib/oauth.ts:
```diff
@@ -1,26 +1,28 @@
 		const userData = await loadUserData();
-		return userData[ TOKEN_KEY ] ?? null;
+		const parsed = authTokenSchema.parse( userData.authToken );
+		return parsed;
```
```suggestion
		return authTokenSchema.parse( userData.authToken );
```

ashfame commented on src/lib/oauth.ts:
```diff
@@ -31,7 +33,7 @@ async function storeToken( tokens: StoredToken ) {
 		await saveUserData( {
 			...userData,
-			[ TOKEN_KEY ]: tokens,
+			authToken: tokens,
```
I believe it's a single token? If so, should rename it to singular? `tokens` to `token`

ashfame commented on src/lib/oauth.ts:
```diff
@@ -67,32 +66,37 @@ export async function isAuthenticated(): Promise< boolean > {
 }
 
-export async function handleAuthCallback( hash: string ): Promise< Error | StoredToken > {
+async function handleAuthCallback( hash: string ): Promise< StoredToken > {
```
👌

ashfame reviewed (approved):
Nice work! I noticed 3 tests that were skipping are also resolved now. 🎆

ashfame commented on src/lib/oauth.ts:
```diff
@@ -67,32 +66,37 @@ export async function isAuthenticated(): Promise< boolean > {
 	try {
-		response = await new wpcom( accessToken ).req.get( '/me?fields=ID,email,display_name' );
+		const response = await new wpcom( accessToken ).req.get( '/me?fields=ID,email,display_name' );
+
```
@fredrikekelund I guess here would be the new point to specifically log it to sentry if `ID` is ever returned empty from api endpoint unless whatever zod will throw would capture that info?

fredrikekelund replied:
We already have a `Sentry.captureException` call in the catch clause, and the nice thing about zod's `parse` function is that it includes a clear explanation of the validation error in the error **message**. Here's an example:
```
[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "null",
    "path": [
      "accessToken"
    ],
    "message": "Expected string, received null"
  }
]
```

ashfame commented on src/components/auth-provider.tsx:
```diff
@@ -48,8 +48,8 @@ const AuthProvider: React.FC< AuthProviderProps > = ( { children } ) => {
 		setClient( createWpcomClient( token.accessToken, locale ) );
 		if ( token.id || token.email || token.displayName ) {
```
This `if` check is now redundant?

ashfame replied:
And `if (error)` seems to be the best place to show an error message to the user?

fredrikekelund replied:
Makes sense 👍 I've added an error modal here, inspired by https://github.com/Automattic/studio/pull/856

fredrikekelund committed:
Merge branch 'trunk' into f26d/validate-auth-token

fredrikekelund committed:
Address review feedback

fredrikekelund committed:
Display error modal when login process fails

fredrikekelund commented:
Thanks for the speedy and diligent review, @ashfame 🙏 I've addressed your feedback and added an "all-purpose" error modal to be displayed if something goes wrong in `handleAuthCallback`.
